### PR TITLE
Add listener task to cancel streaming response in case of an early disconnect

### PR DESCRIPTION
**Pull request by Kamforka**
Related to #20

This one is to fix an ongoing stream to an already disconnected client.

All the existing tests are passed, however I'd like to add one to cover a scenario when a `"http.disconnect"` message is sent via the `receive` channel, @tomchristie maybe you could suggest some sort of setup for such a test?

### CODE REVIEW

**Comment by lovelydinosaur** [comment]
Okay, this is looking decent yup!

I happened to inroduce a [`run_until_first_complete`](https://github.com/encode/starlette/blob/dd764eced233df9356fa3a0a04a2754263fc54c8/starlette/concurrency.py#L14) since I was looking at [a similar pattern for working with websockets](https://github.com/encode/broadcaster/blob/3e8a8bc1dca9c31c33562e8a8802aa089e7688f7/example.py#L20).

I reckon we ought to use that here, and have `.listen_for_disconnect` and `.stream_response_body` methods. That'd cover us in any wierd broken cases where disconnect is never sent for some reason.

**Comment by Kamforka** [comment]
@tomchristie Sure, I'll adjust the code this evening.

Just to be clear, you meant something like this:

```python3
await run_until_first_complete(self.stream_response(send), self.listen_for_disconnect(receive))
```

Right?

**Comment by lovelydinosaur** [comment]
@Kamforka Yup, I'll be along these lines...

```python
await run_until_first_complete(
    (self.stream_response, {"send": send}),
    (self.listen_for_disconnect, {"receive": receive})
)

**Comment by Kamforka** [comment]
@tomchristie added the changes.

However I have one observation. Is there any particular reason for the signature of `run_until_first_complete`?

```python3
async def run_until_first_complete(*args: typing.Tuple[typing.Callable, dict]) -> None
```

I think the following signature would be more intuitive:
```python3
async def run_until_first_complete(tasks: typing.Sequence[typing.Awaitable]) -> None:
```
> Note: `typing.Awaitable` might not be the best type here but you get my point.

So in our particular case one could call it like:
```python3
await run_until_first_complete(
    [self.stream_response(send), self.listen_for_disconnect(receive)]
)
```

As far as I can tell the functionality is the same, however we get less boilerplate with better typing and auto-completion experience. What do you think?

File: starlette/responses.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import asyncio
```
Comment (lovelydinosaur) [inline starlette/responses.py]: Looks like we can drop this now, right?

Comment (Kamforka) [reply starlette/responses.py]: Oh yeah

**Comment by lovelydinosaur** [comment]
Good suggestion yup, although I’m also keeping half an eye here on trio compatibility, which it wouldn’t play so well with.

**Comment by Kamforka** [comment]
What do you mean by trio compatibility?

**Comment by lovelydinosaur** [comment]
> What do you mean by trio compatibility?

We might want a future version of Starlette to support being run on Trio https://trio.readthedocs.io/en/stable/ as an alternative to Asyncio. The API that Trio uses for starting tasks doesn't allow passing instantiated coroutines (for some good reasons) so keeping this constraint in our own codebase would make it easier for us support Trio at some point.

**Comment by lovelydinosaur** [comment]
And, great stuff, thank you!

**Comment by Kamforka** [comment]
> 
> 
> > What do you mean by trio compatibility?
> 
> We might want a future version of Starlette to support being run on Trio https://trio.readthedocs.io/en/stable/ as an alternative to Asyncio. The API that Trio uses for starting tasks doesn't allow passing instantiated coroutines (for some good reasons) so keeping this constraint in our own codebase would make it easier for us support Trio at some point.

Got it, thanks for the clarification!

**Pull request merged**